beacon_node/beacon_chain/src/fork_choice/checkpoint_manager.rs

```diff
@@ -0,0 +1,341 @@
+#[derive(PartialEq, Clone, Default, Encode, Decode)]
+struct BalancesCache {
+    items: Vec<CacheItem>,
+}
+
+impl BalancesCache {
+    /// Inspect the given `state` and determine the root of the block at the first slot of
+    /// `state.current_epoch`. If there is not already some entry for the given block root, then
+    /// add `state.balances` to the cache.
+    pub fn process_state<E: EthSpec>(
+        &mut self,
+        block_root: Hash256,
+        state: &BeaconState<E>,
+    ) -> Result<(), Error> {
+        // We are only interested in balances from states that are at the start of an epoch,
+        // because this is where the `current_justified_checkpoint.root` will point.
+        if !Self::is_first_block_in_epoch(block_root, state)? {
+            return Ok(());
+        }
+
+        let epoch_boundary_slot = state.current_epoch().start_slot(E::slots_per_epoch());
+        let epoch_boundary_root = if epoch_boundary_slot == state.slot {
+            block_root
+        } else {
+            // This call remains sensible as long as `state.block_roots` is larger than a single
+            // epoch.
+            *state.get_block_root(epoch_boundary_slot)?
+        };
+
+        if self.position(epoch_boundary_root).is_none() {
+            let item = CacheItem {
+                block_root: epoch_boundary_root,
+                balances: get_effective_balances(state),
+            };
+
+            if self.items.len() == MAX_BALANCE_CACHE_SIZE {
+                self.items.remove(0);
+            }
+
+            self.items.push(item);
+        }
+
+        Ok(())
+    }
+
+    /// Returns `true` if the given `block_root` is the first/only block to have been processed in
+    /// the epoch of the given `state`.
+    ///
+    /// We can determine if it is the first block by looking back through `state.block_roots` to
+    /// see if there is a block in the current epoch with a different root.
+    fn is_first_block_in_epoch<E: EthSpec>(
+        block_root: Hash256,
+        state: &BeaconState<E>,
+    ) -> Result<bool, Error> {
+        let mut prior_block_found = false;
+
+        for slot in state.current_epoch().slot_iter(E::slots_per_epoch()) {
+            if slot < state.slot {
+                if *state.get_block_root(slot)? != block_root {
+                    prior_block_found = true;
+                    break;
+                }
+            } else {
+                break;
+            }
+        }
+
+        Ok(!prior_block_found)
+    }
+
+    fn position(&self, block_root: Hash256) -> Option<usize> {
+        self.items
+            .iter()
+            .position(|item| item.block_root == block_root)
+    }
+
+    /// Get the balances for the given `block_root`, if any.
+    ///
+    /// If some balances are found, they are removed from the cache.
+    pub fn get(&mut self, block_root: Hash256) -> Option<Vec<u64>> {
+        let i = self.position(block_root)?;
+        Some(self.items.remove(i).balances)
```

Comment:
I'm not sure I follow your description, so perhaps I can respond with my own:

- Every time we call `process_state` with a state that is from the first block in the epoch, we cache it's balances.
- Whenever we process a block that has a new higher CJC, we try the cache for the balances associated with `CJC.root`.
   - Since the cache has a size of `4`, if there are no forks we should get a cache hit as long as there haven't been 4 epochs since justification. If this is not the case (or there are forks washing out the cache) then we load from an on-disk cache.

WRT to the `remove`, I think you're right that we might end up doing unnecessary on-disk reads when we don't update `current`. I'll fix this up in the v0.10.1 update -- in it's current state I don't think it will produce errors, just inefficiency.

Reply:
Ok, I agree with your assessment. I'll leave this convo unresolved so we have a pointer back to it for v0.10.1